jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,427 @@
+                        LabeledPath<V> cycleResult = path.clone();
+                        cycleResult.addVertex(head, graph.getEdgeWeight(graph.getEdge(tail, head)), labelMap.get(head));
+
+                        return new GraphWalk<>(graph, cycleResult.vertices, cycleResult.cost);
+                    }
+
+                    for (E e : graph.outgoingEdgesOf(tail)) {
+                        V currentVertex = graph.getEdgeTarget(e);
+                        // extend the path if the extension is still negative and correctly labeled
+                        double edgeWeight = graph.getEdgeWeight(e);
+                        int currentLabel = labelMap.get(currentVertex);
+                        if (!path.getLabels().contains(currentLabel) && path.getCost() + edgeWeight < 0) {
+                            LabeledPath<V> newPath = path.clone();
+                            newPath.addVertex(currentVertex, edgeWeight, currentLabel);
+
+                            /*
+                             * check if paths are dominated, i.e. if the path is definitely worse than other paths
+                             * and does not have to be considered in the future
+                             */
+                            if (!removeDominatedPathsOfLengthKplus1(newPath, pathsLengthKplus1)) {
+                                if(!removeDominatedPathsOfLengthK(newPath, pathsLengthK)) {
+                                    addPathToPathStructure(pathsLengthKplus1, newPath);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            // update k and the corresponding sets
+            k += 1;
+            pathsLengthK = pathsLengthKplus1;
+            pathsLengthKplus1 = new LinkedHashMap<>();
+        }
+
+        return new GraphWalk<>(graph, new ArrayList<>(lengthBound), Double.MAX_VALUE);
+    }
+
+    /**
+     * Removes all paths that are dominated from all calculated paths of length k + 1.
+     * This is important out of efficiency reasons, otherwise many unnecessary paths may
+     * be considered in further calculations.
+     *
+     * @param path the currently calculated path
+     * @param pathsLengthKplus1 all before calculated paths of length k + 1
+     *
+     * @return whether <code>path</code> was removed in the procedure
+     */
+    private boolean removeDominatedPathsOfLengthKplus1(LabeledPath<V> path, Map<PathSetKey<V>, TreeSet<LabeledPath<V>>> pathsLengthKplus1) {
+        boolean removePath = false;
+
+        TreeSet<LabeledPath<V>> pathsToCheck = pathsLengthKplus1.get(PathSetKey.of(path.getHead(), path.getTail(), path.getLabels()));
+        if(pathsToCheck != null) {
+            // simulates LabeledPath.dominates(LabeledPath) by using the index structure
+            for (Iterator<LabeledPath<V>> it = pathsToCheck.descendingIterator(); it.hasNext(); ) {
```

Comment:
Now that we have the comparison purely on cost, can you think of any more efficient way to do this using the TreeSet interface instead of iteration?  (I don't know the answer since I'm not sure whether it's possible to efficiently remove the entire tail of the tree.)

Reply:
The question is whether `Subset.clear()` runs in O(1). I cannot find anything to this question. Then, it is possible to call `tailset(E)` and call `clear()` on it.